File: platform/sdk/conformance/secret-hygiene.test.ts
import { describe, it, expect } from "@jest/globals";
import { SecretBytes, PlatformSdkError, IdempotencyKey, ProtocolAmount, protocolAmount, idempotencyKey } from "@sidiora/layerx-sdk";

describe("SecretBytes hygiene", () => {
  it("redacts toString", () => {
    const secret = new SecretBytes(new Uint8Array([1, 2, 3, 4]));
    expect(secret.toString()).toBe("[REDACTED]");
  });

  it("redacts toJSON", () => {
    const secret = new SecretBytes(new Uint8Array([1, 2, 3, 4]));
    const json = JSON.stringify({ key: secret });
    expect(json).not.toContain("1");
    expect(json).not.toContain("2");
    expect(json).toContain("[REDACTED]");
  });

  it("zeroizes on destroy", () => {
    const secret = new SecretBytes(new Uint8Array([42, 43, 44]));
    let captured: Uint8Array | undefined;
    secret.withBytes((bytes) => {
      captured = new Uint8Array(bytes);
    });
    expect(captured![0]).toBe(42);
    secret.destroy();
    expect(() => secret.withBytes(() => {})).toThrow(PlatformSdkError);
  });

  it("refuses empty input", () => {
    expect(() => new SecretBytes(new Uint8Array([]))).toThrow(PlatformSdkError);
  });

  it("never exposes material through error serialization", () => {
    const secret = new SecretBytes(new Uint8Array([0xde, 0xad, 0xbe, 0xef]));
    try {
      secret.destroy();
      secret.withBytes(() => {});
    } catch (error: unknown) {
      if (error instanceof PlatformSdkError) {
        const serialized = JSON.stringify(error.toJSON());
        expect(serialized).not.toContain("de");
        expect(serialized).not.toContain("ad");
        expect(serialized).not.toContain("be");
        expect(serialized).not.toContain("ef");
        expect(serialized).not.toContain("dead");
        expect(serialized).not.toContain("beef");
      }
    }
  });

  it("never logs key material when stringified in structured logging context", () => {
    const secret = new SecretBytes(new Uint8Array([0x01, 0x02, 0xff]));
    const logPayload = JSON.stringify({ operation: "sign", key: secret });
    expect(logPayload).toContain("sign");
    expect(logPayload).toContain("[REDACTED]");
    expect(logPayload).not.toContain("01");
    expect(logPayload).not.toContain("02");
    expect(logPayload).not.toContain("ff");
  });
});

describe("IdempotencyKey hygiene", () => {
  it("constructs valid keys", () => {
    const key = idempotencyKey("valid-key-123");
    expect(key).toBeTruthy();
  });

  it("refuses empty keys", () => {
    expect(() => idempotencyKey("")).toThrow(PlatformSdkError);
  });

  it("refuses overlong keys", () => {
    const overlong = "a".repeat(256);
    expect(() => idempotencyKey(overlong)).toThrow(PlatformSdkError);
  });

  it("refuses NUL-containing keys", () => {
    expect(() => idempotencyKey("has\0null")).toThrow(PlatformSdkError);
  });

  it("never leaks key material through error serialization", () => {
    try {
      idempotencyKey("");
    } catch (error: unknown) {
      if (error instanceof PlatformSdkError) {
        const serialized = JSON.stringify(error.toJSON());
        expect(serialized).not.toContain("idempotency");
        expect(serialized).toContain("invalid-argument");
      }
    }
  });
});

describe("ProtocolAmount hygiene", () => {
  it("constructs integer amounts", () => {
    const amount = protocolAmount(12345n);
    expect(amount).toBe(12345n);
  });

  it("parses decimal strings", () => {
    const amount = protocolAmount("67890");
    expect(amount).toBe(67890n);
  });

  it("refuses negative amounts", () => {
    expect(() => protocolAmount(-1n)).toThrow(PlatformSdkError);
  });

  it("refuses amounts exceeding u128", () => {
    const tooLarge = 340282366920938463463374607431768211456n;
    expect(() => protocolAmount(tooLarge)).toThrow(PlatformSdkError);
  });

  it("refuses floating-point representation", () => {
    expect(() => protocolAmount("123.45")).toThrow(PlatformSdkError);
  });

  it("refuses scientific notation", () => {
    expect(() => protocolAmount("1e10")).toThrow(PlatformSdkError);
  });

  it("makes floating-point amounts structurally impossible", () => {
    const amount = protocolAmount(100n);
    expect(typeof amount).toBe("bigint");
  });
});

describe("Error hygiene", () => {
  it("never includes request details in error messages", () => {
    const error = new PlatformSdkError({
      code: "transport-failure",
      retry: "safe",
      requestId: "req-secret-12345",
    });
    expect(error.message).not.toContain("req-secret-12345");
    expect(error.message).toBe("The request could not reach the service.");
  });

  it("serializes only safe machine codes", () => {
    const error = new PlatformSdkError({
      code: "capability-refusal",
      retry: "never",
      protocolResultCode: 4001,
    });
    const serialized = error.toJSON();
    expect(serialized.code).toBe("capability-refusal");
    expect(serialized.retry).toBe("never");
    expect(serialized.protocolResultCode).toBe(4001);
  });

  it("never includes session tokens in serialized errors", () => {
    const error = new PlatformSdkError({
      code: "deadline",
      retry: "safe",
      requestId: "token-Bearer-abc123",
    });
    const serialized = JSON.stringify(error.toJSON());
    expect(serialized).toContain("deadline");
    expect(serialized).toContain("token-Bearer-abc123");
  });
});

File: platform/sdk/conformance/secret-hygiene.test.py
import json
import pytest
from layerx_sdk import (
    SecretBytes,
    PlatformSdkError,
    SdkErrorCode,
    IdempotencyKey,
    ProtocolAmount,
)


class TestSecretBytesHygiene:
    def test_redacts_str(self):
        secret = SecretBytes(b"\x01\x02\x03\x04")
        assert str(secret) == "[REDACTED]"

    def test_redacts_repr(self):
        secret = SecretBytes(b"\x01\x02\x03\x04")
        assert repr(secret) == "SecretBytes([REDACTED])"

    def test_prevents_serialization(self):
        secret = SecretBytes(b"\x42\x43\x44")
        with pytest.raises(TypeError, match="cannot be serialised"):
            import pickle
            pickle.dumps(secret)

    def test_zeroizes_on_del(self):
        secret = SecretBytes(b"\x2a\x2b\x2c")
        captured = None
        def capture(view):
            nonlocal captured
            captured = bytes(view)
        secret.use(capture)
        assert captured[0] == 0x2a
        del secret

    def test_refuses_empty_input(self):
        with pytest.raises(PlatformSdkError) as exc_info:
            SecretBytes(b"")
        assert exc_info.value.code == SdkErrorCode.INVALID_ARGUMENT

    def test_never_exposes_material_through_error_serialization(self):
        secret = SecretBytes(b"\xde\xad\xbe\xef")
        secret.destroy()
        try:
            secret.use(lambda v: None)
        except PlatformSdkError as error:
            serialized = json.dumps(list(error.to_dict().values()))
            assert "de" not in serialized
            assert "ad" not in serialized
            assert "be" not in serialized
            assert "ef" not in serialized
            assert "dead" not in serialized
            assert "beef" not in serialized

    def test_never_logs_key_material_when_stringified(self):
        secret = SecretBytes(b"\x01\x02\xff")
        log_payload = json.dumps({"operation": "sign", "key": str(secret)})
        assert "sign" in log_payload
        assert "[REDACTED]" in log_payload
        assert "01" not in log_payload
        assert "02" not in log_payload
        assert "ff" not in log_payload


class TestIdempotencyKeyHygiene:
    def test_constructs_valid_keys(self):
        key = IdempotencyKey("valid-key-123")
        assert key is not None

    def test_refuses_empty_keys(self):
        with pytest.raises(PlatformSdkError) as exc_info:
            IdempotencyKey("")
        assert exc_info.value.code == SdkErrorCode.INVALID_ARGUMENT

    def test_refuses_overlong_keys(self):
        overlong = "a" * 256
        with pytest.raises(PlatformSdkError):
            IdempotencyKey(overlong)

    def test_refuses_nul_containing_keys(self):
        with pytest.raises(PlatformSdkError):
            IdempotencyKey("has\0null")

    def test_never_leaks_key_material_through_error_serialization(self):
        try:
            IdempotencyKey("")
        except PlatformSdkError as error:
            serialized = json.dumps(error.to_dict())
            assert "invalid-argument" in serialized


class TestProtocolAmountHygiene:
    def test_constructs_integer_amounts(self):
        amount = ProtocolAmount(12345)
        assert amount == 12345

    def test_parses_decimal_strings(self):
        amount = ProtocolAmount("67890")
        assert amount == 67890

    def test_refuses_negative_amounts(self):
        with pytest.raises(PlatformSdkError):
            ProtocolAmount(-1)

    def test_refuses_amounts_exceeding_u128(self):
        too_large = 340282366920938463463374607431768211456
        with pytest.raises(PlatformSdkError):
            ProtocolAmount(too_large)

    def test_refuses_floating_point_representation(self):
        with pytest.raises(PlatformSdkError):
            ProtocolAmount("123.45")

    def test_refuses_scientific_notation(self):
        with pytest.raises(PlatformSdkError):
            ProtocolAmount("1e10")

    def test_makes_floating_point_amounts_structurally_impossible(self):
        amount = ProtocolAmount(100)
        assert isinstance(amount, int)

    def test_refuses_bool_input(self):
        with pytest.raises(PlatformSdkError):
            ProtocolAmount(True)


class TestErrorHygiene:
    def test_never_includes_request_details_in_error_messages(self):
        error = PlatformSdkError(
            SdkErrorCode.TRANSPORT_FAILURE,
            "safe",
            request_id="req-secret-12345",
        )
        assert "req-secret-12345" not in str(error)
        assert str(error) == "The request could not reach the service."

    def test_serializes_only_safe_machine_codes(self):
        error = PlatformSdkError(
            SdkErrorCode.CAPABILITY_REFUSAL,
            "never",
            protocol_result_code=4001,
        )
        serialized = error.to_dict()
        assert serialized["code"] == "capability-refusal"
        assert serialized["retry"] == "never"
        assert serialized["protocol_result_code"] == 4001

    def test_never_includes_session_tokens_in_serialized_errors(self):
        error = PlatformSdkError(
            SdkErrorCode.DEADLINE,
            "safe",
            request_id="token-Bearer-abc123",
        )
        serialized = json.dumps(error.to_dict())
        assert "deadline" in serialized
        assert "token-Bearer-abc123" in serialized
